### PR TITLE
User impl

### PR DESCRIPTION
**Pull request by Ketch92**

### CODE REVIEW

File: src/main/java/core/util/AuthenticationUtils.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+import core.model.exception.InitializationException;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.security.SecureRandom;
+
+public class AuthenticationUtils {
+    public static String hashPassword(String password, byte[] salt) {
+        StringBuilder hashedPassword = new StringBuilder();
+        try {
+            MessageDigest messageDigest = MessageDigest.getInstance("SHA-512");
```
Comment (gigbat) [inline src/main/java/core/util/AuthenticationUtils.java]: here magic number)

**Review by gigbat — APPROVED** [review]

File: src/main/java/core/Main.java [hunk 2]
```diff
@@ -4,27 +4,33 @@
...
 public class Main {
-    private static Injector injector = Injector.getInstance(Main.class.getPackageName());
+    private static final Injector injector
+            = Injector.getInstance(Main.class.getPackageName());
+    private static final MovieService movieService
+            = (MovieService) injector.getInstance(MovieService.class);
+    private static final CinemaHallService cinemaHallService
+            = (CinemaHallService) injector.getInstance(CinemaHallService.class);
+    private static final AuthenticationService authenticationService =
+            (AuthenticationService) injector.getInstance(AuthenticationService.class);
```
Comment (KseniiaMakarova) [inline src/main/java/core/Main.java:28]: let's add MovieSessionService here as well

File: src/main/java/core/Main.java [hunk 3]
```diff
@@ -50,5 +56,22 @@ public static void main(String[] args) {
                 .forEach(System.out::println);
     
         System.out.println("\n=============================\n");
+        
+        User user = new User();
+        user.setEmail("user.email@email.com");
+        user.setPassword("strong_password");
```
Comment (KseniiaMakarova) [inline src/main/java/core/Main.java]: it is redundant to set these values to the user here

File: src/main/java/core/Main.java [hunk 4]
```diff
@@ -50,5 +56,22 @@ public static void main(String[] args) {
...
+        
+        User anotherUser = new User();
+        String anotherUserEmail = "another@email.com";
+        String anotherUserPsw = "ghadghfdas";
+        anotherUser.setEmail(anotherUserEmail);
+        anotherUser.setPassword(anotherUserPsw);
```
Comment (KseniiaMakarova) [inline src/main/java/core/Main.java]: same

File: src/main/java/core/model/User.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+        this.salt = salt;
+    }
+    
+    @Override
+    public String toString() {
+        final StringBuffer sb = new StringBuffer("User{");
+        sb.append("id=").append(id);
+        sb.append(", email='").append(email).append('\'');
+        sb.append('}');
+        return sb.toString();
```
Comment (KseniiaMakarova) [inline src/main/java/core/model/User.java]: let's not use StringBuilder for toString() - it is not practiced
we can use concatenation here

File: src/main/java/core/model/exception/AuthenticationException.java [hunk 6]
```diff
@@ -0,0 +1,12 @@
+package core.model.exception;
+
+public class AuthenticationException extends Exception {
+
```
Comment (KseniiaMakarova) [inline src/main/java/core/model/exception/AuthenticationException.java]: redundant empty line

File: src/main/java/core/service/impl/AuthenticationServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+    private static final String AUTHENTICATION_ERROR_MESSAGE
+            = "User with given email and/or password wasn't found";
+    @Inject
+    private UserService userService;
+    
+    @Override
+    public User login(String email, String password) throws AuthenticationException {
+        User user = userService.findByEmail(email)
+                .orElseThrow(() -> new AuthenticationException(AUTHENTICATION_ERROR_MESSAGE));
+        if (AuthenticationUtils.hashPassword(password,
```
Comment (KseniiaMakarova) [inline src/main/java/core/service/impl/AuthenticationServiceImpl.java]: let's combine all conditions in one `if` and throw the exception only once

File: src/main/java/core/service/impl/AuthenticationServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+                user.getSalt()).equalsIgnoreCase(user.getPassword())) {
```
Comment (KseniiaMakarova) [inline src/main/java/core/service/impl/AuthenticationServiceImpl.java]: why not just equals()? passwords should be exactly the same

File: src/main/java/core/util/AuthenticationUtils.java [hunk 9]
```diff
@@ -0,0 +1,32 @@
+package core.util;
+
+import core.model.exception.InitializationException;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.security.SecureRandom;
+
+public class AuthenticationUtils {
```
Comment (KseniiaMakarova) [inline src/main/java/core/util/AuthenticationUtils.java]:
```suggestion
public class AuthenticationUtil {
```

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

File: src/main/java/core/Main.java (hunk 2 above)
Comment (Ketch92) [reply src/main/java/core/Main.java:28]: ok

File: src/main/java/core/Main.java (hunk 3 above)
Comment (Ketch92) [reply src/main/java/core/Main.java]: sure

File: src/main/java/core/Main.java (hunk 4 above)
Comment (Ketch92) [reply src/main/java/core/Main.java]: removed

File: src/main/java/core/service/impl/AuthenticationServiceImpl.java (hunk 8 above)
Comment (Ketch92) [reply src/main/java/core/service/impl/AuthenticationServiceImpl.java]: Ohh... I had a problem there, but it equals wasn't the reason

File: src/main/java/core/util/AuthenticationUtils.java (hunk 9 above)
Comment (Ketch92) [reply src/main/java/core/util/AuthenticationUtils.java]: sure

File: src/main/java/core/Main.java [hunk 10]
```diff
@@ -14,28 +14,30 @@
 import java.time.LocalDateTime;
 
 public class Main {
-    private static final Injector injector
+    private static final Injector INJECTOR
```
Comment (KseniiaMakarova) [inline src/main/java/core/Main.java]: although this is static and final - this is not a magic number or string, or something like that
let's change the name back to `lowercase`

File: src/main/java/core/Main.java [hunk 11]
```diff
@@ -14,28 +14,30 @@
...
-    private static final AuthenticationService authenticationService =
-            (AuthenticationService) injector.getInstance(AuthenticationService.class);
+    private static final MovieService MOVIE_SERVICE
+            = (MovieService) INJECTOR.getInstance(MovieService.class);
+    private static final CinemaHallService CINEMA_HALL_SERVICE
+            = (CinemaHallService) INJECTOR.getInstance(CinemaHallService.class);
+    private static final MovieSessionService SESSION_SERVICE
+            = (MovieSessionService) INJECTOR.getInstance(MovieSessionService.class);
+    private static final AuthenticationService AUTHENTICATION_SERVICE =
+            (AuthenticationService) INJECTOR.getInstance(AuthenticationService.class);
```
Comment (KseniiaMakarova) [inline src/main/java/core/Main.java]: same here

File: src/main/java/core/model/User.java [hunk 12]
```diff
@@ -54,11 +54,12 @@ public void setSalt(byte[] salt) {
...
-        final StringBuffer sb = new StringBuffer("User{");
-        sb.append("id=").append(id);
-        sb.append(", email='").append(email).append('\'');
-        sb.append('}');
-        return sb.toString();
+        return "User{"
+               + "id=" + id
+               + ", email='" + email + '\''
+               + ", password='" + password + '\''
+               + ", salt=" + Arrays.toString(salt)
```
Comment (KseniiaMakarova) [inline src/main/java/core/model/User.java]: let's not make secure data visible

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

File: src/main/java/core/Main.java (hunk 11 above)
Comment (Ketch92) [reply src/main/java/core/Main.java]: Done

File: src/main/java/core/Main.java (hunk 10 above)
Comment (Ketch92) [reply src/main/java/core/Main.java]: Done

File: src/main/java/core/model/User.java (hunk 12 above)
Comment (Ketch92) [reply src/main/java/core/model/User.java]: Forgotten about it

**Review by KseniiaMakarova — APPROVED** [review]

**Pull request merged**